constructr-coordination/src/main/scala/de/heikoseeberger/constructr/coordination/ConsulCoordination.scala

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright 2015 Heiko Seeberger
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package de.heikoseeberger.constructr.coordination
+
+import akka.http.scaladsl.client.RequestBuilding.{ Get, Put }
+import akka.http.scaladsl.model.StatusCodes.{ NotFound, OK }
+import akka.http.scaladsl.model.{ HttpRequest, HttpResponse, ResponseEntity, Uri }
+import akka.http.scaladsl.unmarshalling.Unmarshal
+import akka.stream.Materializer
+
+import scala.concurrent.duration.Duration
+import scala.concurrent.{ ExecutionContext, Future }
+
+final class ConsulCoordination(prefix: String, clusterName: String, host: String, port: Int, send: HttpRequest => Future[HttpResponse])
+    extends Coordination(prefix, clusterName, host, port, send) {
+  import Coordination._
+
+  private val v1Uri = Uri(s"http://$host:$port/v1")
+
+  private val kvUri = Uri(s"http://$host:$port/v1/kv")
+
+  private val baseUri = kvUri.withPath(kvUri.path / "constructr" / prefix / clusterName)
+
+  private val nodesUri = baseUri.withPath(baseUri.path / "nodes")
+
+  private val createSessionUri = v1Uri.withPath(v1Uri.path / "session" / "create")
+
+  //private def renewSessionUri(sessionId: String) = kvUri.withPath(kvUri.path / "session" / "renew" / sessionId)
```

Comment:
To begin with, we'd need `addSelf` to return somehow the `sessionId` of the session that is linked to the new key. Following your design for `LockResult` should be something like this:
```scala
sealed trait AddingSelfResult
object AddingSelfResult {
  case class Success(sessionId: String) extends AddingSelfResult
  case object Failure extends AddingSelfResult
}
```
However, this design is not general enough as the `session` concept applies to Consul but doesn't apply to etcd. So, it could be something like this:
```scala
sealed trait AddingSelfResult[T]
object AddingSelfResult {
  case class Success[T](result: T) extends AddingSelfResult[T]
  case object Failure extends AddingSelfResult[Nothing]
}
```
Apart from that, we should be able to keep track the `sessionId` in the FSM State, so we could extend the `Data[A]` type constructor to incorporate that info. Could be sth as follows:
```scala
case class Data[A, B](nodes: List[A], coordinationRetriesLeft: Int, AddingSelfResult: B)
```
And, yes we should probably come back to the previous design where AddingSelf and Refreshing were separated.

WDYT?